File: app/src/main/java/com/eveningoutpost/dexdrip/tidepool/BaseElement.java
package com.eveningoutpost.dexdrip.tidepool;

import com.google.gson.annotations.Expose;

/**
 * jamorham
 * <p>
 * common element base
 */

public abstract class BaseElement {
    @Expose
    public String deviceTime;
    @Expose
    public String time;
    @Expose
    public int timezoneOffset;
    @Expose
    public String type;
    @Expose
    public Origin origin;


    BaseElement populate(final long timestamp, final String uuid) {
        deviceTime = DateUtil.toFormatNoZone(timestamp);
        time = DateUtil.toFormatAsUTC(timestamp);
        timezoneOffset = DateUtil.getTimeZoneOffsetMinutes(timestamp); // TODO
        origin = new Origin(uuid);
        return this;
    }

    public class Origin {
        @Expose
        String id;

        Origin(String id) {
            this.id = id;
        }
    }
}

File: plugins/sync/src/androidMain/kotlin/app/aaps/plugins/sync/tidepool/elements/BaseElement.kt
package app.aaps.plugins.sync.tidepool.elements

import app.aaps.core.interfaces.utils.DateUtil
import com.google.gson.annotations.Expose

open class BaseElement(timestamp: Long, uuid: String, dateUtil: DateUtil) {

    @Expose
    var deviceTime: String = ""

    @Expose
    var time: String = ""

    @Expose
    var timezoneOffset: Int = 0

    @Expose
    var type: String? = null

    @Expose
    var origin: Origin? = null

    init {
        deviceTime = dateUtil.toISONoZone(timestamp)
        time = dateUtil.toISOAsUTC(timestamp)
        timezoneOffset = dateUtil.getTimeZoneOffsetMinutes(timestamp) // TODO
        origin = Origin(uuid)
    }

    inner class Origin internal constructor(
        @field:Expose
        internal var id: String
    )
}